File: android/app/src/main/java/com/noc/app/core/crypto/Wire.kt
package com.noc.app.core.crypto

import java.io.ByteArrayOutputStream
import java.nio.ByteBuffer
import java.security.MessageDigest
import java.util.Base64
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/** Primitivas de codificação do protocolo Noc v1 (ver protocol/PROTOCOL.md). */
object Wire {
    const val FRAME_CLIENT_HELLO: Byte = 0x01
    const val FRAME_SERVER_HELLO: Byte = 0x02
    const val FRAME_ENCRYPTED: Byte = 0x10
    const val FRAME_ERROR: Byte = 0x7F
    const val PROTOCOL_VERSION: Byte = 1

    val LABEL_TRANSCRIPT = "NOC/1/transcript".toByteArray(Charsets.US_ASCII)
    val LABEL_SERVER = "NOC/1/server".toByteArray(Charsets.US_ASCII)
    val LABEL_CLIENT = "NOC/1/client".toByteArray(Charsets.US_ASCII)
    val LABEL_PAIR = "NOC/1/pair".toByteArray(Charsets.US_ASCII)
    val INFO_KEYS = "NOC/1/keys".toByteArray(Charsets.US_ASCII)
    val INFO_SAS = "NOC/1/sas".toByteArray(Charsets.US_ASCII)

    fun lp(data: ByteArray): ByteArray =
        ByteBuffer.allocate(4 + data.size).putInt(data.size).put(data).array()

    fun concat(vararg parts: ByteArray): ByteArray {
        val out = ByteArrayOutputStream(parts.sumOf { it.size })
        parts.forEach { out.write(it) }
        return out.toByteArray()
    }

    /** Lê um campo lp; devolve (valor, próximo offset). */
    fun readLp(data: ByteArray, offset: Int, maxLen: Int = 4096): Pair<ByteArray, Int> {
        if (data.size - offset < 4) throw ProtocolException("truncated")
        val len = ByteBuffer.wrap(data, offset, 4).int
        if (len < 0 || len > maxLen || data.size - offset - 4 < len) throw ProtocolException("truncated")
        return data.copyOfRange(offset + 4, offset + 4 + len) to offset + 4 + len
    }

    private const val B32 = "abcdefghijklmnopqrstuvwxyz234567"

    fun base32(data: ByteArray): String {
        val sb = StringBuilder((data.size * 8 + 4) / 5)
        var bits = 0
        var value = 0
        for (b in data) {
            value = (value shl 8) or (b.toInt() and 0xff)
            bits += 8
            while (bits >= 5) {
                sb.append(B32[(value shr (bits - 5)) and 31])
                bits -= 5
            }
        }
        if (bits > 0) sb.append(B32[(value shl (5 - bits)) and 31])
        return sb.toString()
    }

    fun sha256(vararg parts: ByteArray): ByteArray {
        val md = MessageDigest.getInstance("SHA-256")
        parts.forEach { md.update(it) }
        return md.digest()
    }

    fun idFromSpki(spki: ByteArray): String = base32(sha256(spki)).substring(0, 26)

    fun hmacSha256(key: ByteArray, data: ByteArray): ByteArray {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(key, "HmacSHA256"))
        return mac.doFinal(data)
    }

    /** HKDF-SHA256 (RFC 5869). */
    fun hkdf(ikm: ByteArray, salt: ByteArray, info: ByteArray, length: Int): ByteArray {
        val prk = hmacSha256(if (salt.isEmpty()) ByteArray(32) else salt, ikm)
        val out = ByteArrayOutputStream(length)
        var t = ByteArray(0)
        var counter = 1
        while (out.size() < length) {
            t = hmacSha256(prk, concat(t, info, byteArrayOf(counter.toByte())))
            out.write(t)
            counter++
        }
        return out.toByteArray().copyOf(length)
    }

    fun b64(data: ByteArray): String = Base64.getEncoder().encodeToString(data)
    fun unb64(s: String): ByteArray = Base64.getDecoder().decode(s)
    fun b64Url(data: ByteArray): String = Base64.getUrlEncoder().withoutPadding().encodeToString(data)
    fun unb64Url(s: String): ByteArray = Base64.getUrlDecoder().decode(s)

    fun hex(data: ByteArray): String = data.joinToString("") { "%02x".format(it) }
    fun unhex(s: String): ByteArray = ByteArray(s.length / 2) { s.substring(it * 2, it * 2 + 2).toInt(16).toByte() }
}

class ProtocolException(val code: String) : Exception(code)

File: companion/src/Noc.Core/Crypto/Wire.cs
using System.Buffers.Binary;
using System.Security.Cryptography;
using System.Text;

namespace Noc.Core.Crypto;

/// <summary>Primitivas de codificação usadas pelo protocolo (ver protocol/PROTOCOL.md).</summary>
public static class Wire
{
    public const byte FrameClientHello = 0x01;
    public const byte FrameServerHello = 0x02;
    public const byte FrameEncrypted = 0x10;
    public const byte FrameError = 0x7F;
    public const byte ProtocolVersion = 1;

    public static readonly byte[] LabelTranscript = Encoding.ASCII.GetBytes("NOC/1/transcript");
    public static readonly byte[] LabelServer = Encoding.ASCII.GetBytes("NOC/1/server");
    public static readonly byte[] LabelClient = Encoding.ASCII.GetBytes("NOC/1/client");
    public static readonly byte[] LabelPair = Encoding.ASCII.GetBytes("NOC/1/pair");
    public static readonly byte[] LabelRelay = Encoding.ASCII.GetBytes("NOC/1/relay");
    public static readonly byte[] InfoKeys = Encoding.ASCII.GetBytes("NOC/1/keys");
    public static readonly byte[] InfoSas = Encoding.ASCII.GetBytes("NOC/1/sas");

    public static byte[] Lp(ReadOnlySpan<byte> data)
    {
        var buf = new byte[4 + data.Length];
        BinaryPrimitives.WriteUInt32BigEndian(buf, (uint)data.Length);
        data.CopyTo(buf.AsSpan(4));
        return buf;
    }

    public static byte[] Concat(params byte[][] parts)
    {
        var total = 0;
        foreach (var p in parts) total += p.Length;
        var buf = new byte[total];
        var o = 0;
        foreach (var p in parts) { p.CopyTo(buf, o); o += p.Length; }
        return buf;
    }

    /// <summary>Lê um campo lp a partir de <paramref name="offset"/>, avançando-o.</summary>
    public static byte[] ReadLp(ReadOnlySpan<byte> data, ref int offset, int maxLen = 4096)
    {
        if (data.Length - offset < 4) throw new ProtocolException("truncated");
        var len = BinaryPrimitives.ReadUInt32BigEndian(data[offset..]);
        offset += 4;
        if (len > maxLen || data.Length - offset < len) throw new ProtocolException("truncated");
        var result = data.Slice(offset, (int)len).ToArray();
        offset += (int)len;
        return result;
    }

    private const string B32 = "abcdefghijklmnopqrstuvwxyz234567";

    public static string Base32(ReadOnlySpan<byte> data)
    {
        var sb = new StringBuilder((data.Length * 8 + 4) / 5);
        int bits = 0, value = 0;
        foreach (var b in data)
        {
            value = (value << 8) | b;
            bits += 8;
            while (bits >= 5)
            {
                sb.Append(B32[(value >> (bits - 5)) & 31]);
                bits -= 5;
            }
        }
        if (bits > 0) sb.Append(B32[(value << (5 - bits)) & 31]);
        return sb.ToString();
    }

    /// <summary>pcId / deviceId: 26 primeiros caracteres de base32(SHA256(SPKI)).</summary>
    public static string IdFromSpki(ReadOnlySpan<byte> spki) => Base32(SHA256.HashData(spki))[..26];

    public static string B64Url(ReadOnlySpan<byte> data) =>
        Convert.ToBase64String(data).TrimEnd('=').Replace('+', '-').Replace('/', '_');

    public static byte[] FromB64Url(string s)
    {
        var t = s.Replace('-', '+').Replace('_', '/');
        switch (t.Length % 4) { case 2: t += "=="; break; case 3: t += "="; break; }
        return Convert.FromBase64String(t);
    }
}

public sealed class ProtocolException(string code) : Exception(code)
{
    public string Code { get; } = code;
}
